crates/core/src/host/v8/de.rs

```diff
@@ -0,0 +1,420 @@
+    // see: https://github.com/v8/v8/blob/14ac92e02cc3db38131a57e75e2392529f405f2f/include/v8.h#L3165-L3171
+    v8::String::new_from_utf8(scope, field.as_ref(), v8::NewStringType::Internalized).unwrap()
+}
+
+/// Extracts a reference `&'s T` from an owned V8 [`Local<'s, T>`].
+///
+/// The lifetime `'s` is that of the [`HandleScope<'s>`].
+/// This ensures that the reference to `T` won't outlive the `HandleScope`.
+fn deref_local<'s, T>(local: Local<'s, T>) -> &'s T {
+    let reference = local.borrow();
+    // SAFETY: Lifetime extend `'0` to `'s`.
+    // This is safe as the returned reference `&'s T`
+    // will not outlive its `HandleScope<'s, _>`,
+    // as both are tied to the lifetime `'s`.
+    unsafe { core::mem::transmute::<&T, &'s T>(reference) }
+}
+
+/// Deserializes a primitive via [`FromValue`].
+macro_rules! deserialize_primitive {
+    ($dmethod:ident, $t:ty) => {
+        fn $dmethod(self) -> Result<$t, Self::Error> {
+            FromValue::from_value(self.input, self.common.scope).map_err(Error::Value)
+        }
+    };
+}
+
+impl<'de, 'a, 's: 'de> de::Deserializer<'de> for Deserializer<'a, 's> {
+    type Error = Error<'s>;
+
+    // Deserialization of primitive types defers to `FromValue`.
+    deserialize_primitive!(deserialize_bool, bool);
+    deserialize_primitive!(deserialize_u8, u8);
+    deserialize_primitive!(deserialize_u16, u16);
+    deserialize_primitive!(deserialize_u32, u32);
+    deserialize_primitive!(deserialize_u64, u64);
+    deserialize_primitive!(deserialize_u128, u128);
+    deserialize_primitive!(deserialize_u256, u256);
+    deserialize_primitive!(deserialize_i8, i8);
+    deserialize_primitive!(deserialize_i16, i16);
+    deserialize_primitive!(deserialize_i32, i32);
+    deserialize_primitive!(deserialize_i64, i64);
+    deserialize_primitive!(deserialize_i128, i128);
+    deserialize_primitive!(deserialize_i256, i256);
+    deserialize_primitive!(deserialize_f64, f64);
+    deserialize_primitive!(deserialize_f32, f32);
+
+    fn deserialize_product<V: ProductVisitor<'de>>(self, visitor: V) -> Result<V::Output, Self::Error> {
+        let object = cast!(
+            self.common.scope,
+            self.input,
+            Object,
+            "object for product type `{}`",
+            visitor.product_name().unwrap_or("<unknown>")
+        )?;
+
+        visitor.visit_named_product(ProductAccess {
+            common: self.common,
+            object,
+            next_value: None,
+            index: 0,
+        })
+    }
+
+    fn deserialize_sum<V: SumVisitor<'de>>(self, visitor: V) -> Result<V::Output, Self::Error> {
+        let scope = &mut *self.common.scope;
+        let sum_name = visitor.sum_name().unwrap_or("<unknown>");
+
+        // We expect a canonical representation of a sum value in JS to be
+        // `{ tag: "foo", value: a_value_for_foo }`
+        // with special convenience for optionals
+        // where we also accept `null`, `undefined` and an object without `tag`.
```

Comment:
It feels mildly footgun-ey that putting a `tag` within an object breaks using the convenience representation, and therefore that `Option<enum { Foo, Bar }>` must be represented as `{ "tag": "Some", "value": { "tag": "Foo" } }`. I think we probably should require that optionals be `null | undefined | bare_value`, never the "regular sum" representation.